test/unit/BPool.t.sol

```diff
@@ -725,21 +726,124 @@ contract BPool_Unit_Finalize is BasePoolTest {
+  }
 
-  function test_Emit_LogCall() private view {}
+  modifier happyPath(Bind_FuzzScenario memory _fuzz) {
+    _assumeHappyPath(_fuzz);
+    _setValues(_fuzz);
+    _;
+  }
+
+  function test_HappyPath(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+  }
+
+  function test_Revert_NotController(
+    address _controller,
+    address _caller,
+    Bind_FuzzScenario memory _fuzz
+  ) public happyPath(_fuzz) {
+    vm.assume(_controller != _caller);
+    bPool.set__controller(_controller);
+
+    vm.prank(_caller);
+    vm.expectRevert('ERR_NOT_CONTROLLER');
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+  }
+
+  function test_Revert_IsBound(uint256 _tokenIndex, Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    vm.assume(_fuzz.previousTokensAmount > 0);
+    _tokenIndex = bound(_tokenIndex, 0, _fuzz.previousTokens.length - 1);
+
+    vm.expectRevert('ERR_IS_BOUND');
+    bPool.bind(_fuzz.previousTokens[_tokenIndex], _fuzz.balance, _fuzz.denorm);
+  }
+
+  function test_Revert_Finalized(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    _setFinalize(true);
+
+    vm.expectRevert('ERR_IS_FINALIZED');
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+  }
+
+  function test_Revert_MaxPoolTokens(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    address[] memory _tokens = _setRandomTokens(MAX_BOUND_TOKENS);
+    for (uint256 i = 0; i < _tokens.length; i++) {
+      vm.assume(_fuzz.token != _tokens[i]);
+    }
+
+    vm.expectRevert('ERR_MAX_TOKENS');
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+  }
 
-  function test_Call_Rebind() private view {}
+  function test_Set_Record(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+
+    assertTrue(bPool.isBound(_fuzz.token));
+    assertEq(bPool.call__records(_fuzz.token).index, _fuzz.previousTokensAmount);
+  }
+
+  function test_Set_TokenArray(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+
+    assertEq(bPool.getCurrentTokens().length, _fuzz.previousTokensAmount + 1);
+    assertEq(bPool.getCurrentTokens()[_fuzz.previousTokensAmount], _fuzz.token);
+  }
+
+  function test_Emit_LogCall(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    vm.expectEmit();
+    bytes memory _data = abi.encodeWithSelector(BPool.bind.selector, _fuzz.token, _fuzz.balance, _fuzz.denorm);
+    emit BPool.LOG_CALL(BPool.bind.selector, address(this), _data);
+
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
+  }
+
+  function test_Call_Rebind(Bind_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    // TODO: fix
+    vm.skip(true);
+    vm.expectCall(
+      address(bPool), abi.encodeWithSelector(BPool.rebind.selector, _fuzz.token, _fuzz.balance, _fuzz.denorm)
+    );
+
+    bPool.bind(_fuzz.token, _fuzz.balance, _fuzz.denorm);
```

Comment:
A call shouldn't be expected because what Solidity does when "calling" a `public` function from the same contract is to execute a `JUMP` opcode, as if `rebind` were `internal` or `private`.

In cases like these, the safest way to go is to re-test the whole function being called in the context of the calling one. Alternatively, `Smock Foundry` might bring some useful workarounds.

Reply:
Any reccomendations on how to do it with smock? Checked smock but could not figure out how.
I was thinking about `expectEmit` an event by the `rebind()` or checking that `rebind` does set a variable like `totalWeight`, since testing again the entire function means repeating code.

Reply:
Hmm, I checked `Smock Foundry` and it doesn't seem to currently have a workaround for this, as the `public` function would need to be `internal virtual` for the mock to apply.

I like the approach of expecting an event emitted just by the called function, but such ends up being an extra-programmatic assumption; again, the safest way to go would be to re-test it entirely, even repeating code.